### PR TITLE
Expose inner SGD options in constrained solvers

### LINKED ISSUE #179: Expose `stochastic_gradient_descent`'s keyword-arguments in `penalty_method` and `augmented_lagrangian`
The `penalty_method` and `augmented_lagrangian` minimisation solvers internally use `stochastic_gradient_descent`. It would very likely be useful to users to be able to pass some argument into the outer solve to be passed on to the inner solver.

### PR DESCRIPTION
**Pull request by lntutor**
## Summary

Add an `sgd_kwargs` option to the penalty and augmented Lagrangian solvers so callers can configure the inner stochastic-gradient-descent solve. Internal objective arguments remain protected while options such as learning rate, iteration limit, tolerance, optimiser, callbacks, and logging can be forwarded.

## Tests

```
uv run --extra test pytest tests/test_solvers/test_penalty.py tests/test_solvers/test_augmented_lagrangian.py -q
uv run ruff check src/causalprog/solvers/penalty.py src/causalprog/solvers/augmented_lagrangian.py tests/test_solvers/test_penalty.py tests/test_solvers/test_augmented_lagrangian.py
uv run ruff format --check src/causalprog/solvers/penalty.py src/causalprog/solvers/augmented_lagrangian.py tests/test_solvers/test_penalty.py tests/test_solvers/test_augmented_lagrangian.py
```

Fixes #179

### CODE REVIEW

**Comment by lntutor** [comment]
Current head `ffbd3fb7b970324f3dd0efb901b8859eacf9b058` is mergeable, and there are no unresolved review threads on the current PR state. This looks ready for maintainer review when convenient.

File: src/causalprog/solvers/augmented_lagrangian.py [hunk 1]
```diff
@@ -71,6 +75,11 @@ def minimise(
     """
     if fn_kwargs is None:
         fn_kwargs = {}
+    if sgd_kwargs is None:
+        sgd_kwargs = {}
+    if {"fn_args", "fn_kwargs"} & sgd_kwargs.keys():
```
Comment (willGraham01) [inline src/causalprog/solvers/augmented_lagrangian.py:80]:
```suggestion
    elif {"fn_args", "fn_kwargs"} & sgd_kwargs.keys():
```

Since we just checked and handled the default case. Additionally, the `learning_rate` argument to `stochastic_gradient_descent` is also reserved by us in this method, so this should be added to this list.

Also whilst the set-comparison followed by implicit boolean cast does work, you could also consider the more verbose

```suggestion
    elif "fn_args" in sgd or "fn_kwargs" in sgd:
```

just for readability.

File: src/causalprog/solvers/augmented_lagrangian.py [hunk 2]
```diff
@@ -112,11 +121,10 @@ def is_converged(x: PyTree, dx: PyTree) -> bool:
...
         previous_solution = current_solution
+        inner_sgd_kwargs = {"learning_rate": 1 / mu, **sgd_kwargs}
+        inner_sgd_kwargs["fn_kwargs"] = {"mu": mu, "lamb": lamb}
         current_solution = stochastic_gradient_descent(
-            objective,
-            current_solution,
-            fn_kwargs={"mu": mu, "lamb": lamb},
-            learning_rate=1 / mu,
+            objective, current_solution, **inner_sgd_kwargs
         ).fn_args
```
Comment (willGraham01) [inline src/causalprog/solvers/augmented_lagrangian.py:128]:
```suggestion
current_solution = stochastic_gradient_descent(
            objective,
            current_solution,
            fn_kwargs={"mu": mu, "lamb": lamb},
            learning_rate = 1 / mu,
            **inner_sgd_kwargs,
).fn_args
```

I don't think there's a good reason to pack and then immediately unpack a dictionary here. Plus this has the bonus of making it clear what we are forwarding and what we are managing.

Also, the current implementation silently overrides any user-specified "learning_rate" that might be passed in via `inner_sgd_kwargs`. I think this case should be caught explicitly (and have mentioned this above).

File: src/causalprog/solvers/penalty.py
Comment (willGraham01) [inline src/causalprog/solvers/penalty.py:1]: My comments on this file are just analogies of those on the `augmented_lagrangian.py` file.

File: tests/test_solvers/test_augmented_lagrangian.py [hunk 3]
```diff
@@ -34,6 +35,38 @@ def test_maximise():
     assert jnp.isclose(max_solution.obj_val, 5.0 - min_solution.obj_val)
 
 
+def test_minimise_forwards_sgd_kwargs(monkeypatch):
```
Comment (willGraham01) [inline tests/test_solvers/test_augmented_lagrangian.py:38]: We prefer the use [of `pytest-mock`](https://github.com/pytest-dev/pytest-mock/) in our test suite, as opposed to `monkeypatch`.

This also means that you should be able use `mocker.spy` to record the values that `stochastic_gradient_descent` is called with, rather than needing the current implementation of writing a function that dynamically updates a "global" dictionary. Simply running the `augmented_lagrangian` method for one iteration should be enough to check the arguments are forwarded, without needing a "fake sgd" solver.

File: tests/test_solvers/test_penalty.py
Comment (willGraham01) [inline tests/test_solvers/test_penalty.py:1]: Analogous comments here as on the `test_augmented_lagrangian.py` file.

**Review by willGraham01 — CHANGES_REQUESTED** [review]
Hi @lntutor, thanks for making your first contribution!

I left a few comments below which I'd like to discuss / address before approving this PR. Please feel free to use the threads below to reply.

**Comment by mscroggs** [comment]
Closing as there are conflicts with current main branch. Please re-open if you are able to resolve these
